packages/services/src/contents/index.ts

```diff
@@ -1571,3 +1575,259 @@ namespace Private {
     return extension;
   }
 }
+
+class ContentProviderRegistry implements IContentProviderRegistry {
+  constructor() {
+    this._providers = [];
+    this._drive = undefined;
+  }
+
+  get drive(): Contents.IDrive | undefined {
+    return this._drive;
+  }
+
+  set drive(drive: Contents.IDrive) {
+    this._drive = drive;
+    for (let provider of this._providers) {
+      provider.drive = drive;
+    }
+  }
+
+  register(provider: IContentProvider): void {
+    if (this._drive !== undefined) {
+      provider.drive = this._drive;
+    }
+    this._providers.push(provider);
+  }
+
+  getProvider(filePath: string): IContentProvider | undefined {
+    let ext = filePath.split('.').pop();
+    if (ext == null) ext = '';
+    let score: number = 0;
+    let bestProvider: IContentProvider | undefined = undefined;
+    for (let provider of this._providers) {
+      for (let extension of provider.extensions) {
+        const re = new RegExp(extension.re);
+        if (ext.match(re) !== null) {
+          if (extension.score > score) {
+            bestProvider = provider;
+            score = extension.score;
+          }
+        }
+      }
+    }
+    return bestProvider;
+  }
+
+  private _providers: IContentProvider[];
+  private _drive: Contents.IDrive | undefined;
+}
+
+/**
+ * A content provider using the Jupyter REST API.
+ */
+export class RestContentProvider implements IContentProvider {
+  constructor() {
+    this._apiEndpoint = SERVICE_DRIVE_URL;
+  }
+
+  get drive(): Contents.IDrive {
+    return this._drive;
+  }
+
+  set drive(drive: Contents.IDrive) {
+    this._drive = drive;
+  }
```

Comment:
Neither getter nor setter have side effects, this can be simplified down to:

```suggestion
  drive: Contents.IDrive;
```